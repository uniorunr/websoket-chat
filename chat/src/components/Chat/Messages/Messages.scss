@import "../../../colors/colors";

.messages {
  flex-grow: 1;
  width: 100%;
  display: flex;
  justify-content: center;
  border-left: 2px solid $black-color;
  border-right: 2px solid $black-color;
  font-size: 1.4rem;
  overflow: auto;
  background: $white-color;

  .messages__content {
    display: flex;
    flex-direction: column;
    width: 100%;
    padding: 10px 20px;
    justify-content: flex-start;
    overflow: auto;

    .messages__author {
      width: fit-content;
      margin: 5px 0 0;
      max-width: 85%;
      font-size: 1.4rem;
      color: $black-color;
      overflow-wrap: break-word;
      font-weight: 700;
    }

    .messages__date {
      width: fit-content;
      margin: 0 0 5px 0;
      max-width: 85%;
      font-size: 1rem;
      color: $black-color;
      overflow-wrap: break-word;
      font-style: italic;
    }

    .messages__message {
      width: fit-content;
      max-width: 85%;
      padding: 10px 15px;
      margin: 5px 0;
      background: $message-dark-color;
      color: $white-color;
      border-radius: 10px;
      overflow-wrap: break-word;
    }

    .messages__own {
      width: 100%;
      display: flex;
      flex-direction: column;
      align-self: flex-end;

      .messages__author {
        align-self: flex-end;

      }

      .messages__date {
        align-self: flex-end;
      }

      .messages__message {
        align-self: flex-end;
        background: $message-light-color;
        color: $black-color;
      }
    }
  }
}

