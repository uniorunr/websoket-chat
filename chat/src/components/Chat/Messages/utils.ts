const formatDate = (ms: number) => {
  const date = new Date(ms);
  const hours = date.getHours() >= 10 ? date.getHours() : `0${date.getHours()}`;
  const minutes =
    date.getMinutes() >= 10 ? date.getMinutes() : `0${date.getMinutes()}`;
  const seconds =
    date.getSeconds() >= 10 ? date.getSeconds() : `0${date.getSeconds()}`;
  const day = date.getDate() >= 10 ? date.getDate() : `0${date.getDate()}`;
  const month =
    date.getMonth() >= 10 ? date.getMonth() + 1 : `0${date.getMonth() + 1}`;
  const year = date.getFullYear();

  return `${hours}:${minutes}:${seconds} ${day}-${month}-${year}`;
};

export { formatDate };
