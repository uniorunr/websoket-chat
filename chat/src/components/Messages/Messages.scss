.messages {
  flex-grow: 1;
  width: 100%;
  display: flex;
  justify-content: center;
  border-left: 2px solid black;
  border-right: 2px solid black;
  font-size: 1.4rem;
  overflow: auto;

  .messages__content {
    display: flex;
    flex-direction: column;
    width: 100%;
    padding: 10px 20px;
    justify-content: flex-start;

    .messages__author {
      width: fit-content;
      margin: 10px 0 0;
      max-width: 80%;
      font-size: 1.2rem;
      color: #505050;
    }

    .messages__message {
      width: fit-content;
      max-width: 80%;
      padding: 10px 15px;
      margin: 5px 0;
      background: #232323;
      color: #ffffff;
      border-radius: 10px;
    }
  }
}
