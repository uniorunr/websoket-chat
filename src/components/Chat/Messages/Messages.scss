@import "../../../theme/colors";

.messages {
  flex-grow: 1;
  width: 100%;
  display: flex;
  justify-content: center;
  border-left: 2px solid $black-color;
  border-right: 2px solid $black-color;
  font-size: 1.4rem;
  overflow: auto;
  background: $white-color;

  .messages__content {
    display: flex;
    flex-direction: column;
    width: 100%;
    padding: 10px 20px;
    justify-content: flex-start;
    overflow: auto;

    .messages__author {
      width: fit-content;
      margin: 5px 0 0;
      max-width: 85%;
      font-size: 1.4rem;
      color: $black-color;
      overflow-wrap: break-word;
      font-weight: 700;
    }

    .messages__date {
      width: fit-content;
      margin: 0 0 5px 0;
      max-width: 85%;
      font-size: 1rem;
      color: $black-color;
      overflow-wrap: break-word;
      font-style: italic;
    }

    .messages__message {
      width: fit-content;
      max-width: 85%;
      padding: 10px 15px;
      margin: 5px 0;
      background: $message-dark-color;
      color: $white-color;
      border-radius: 10px;
      overflow-wrap: break-word;
    }

    .messages__own {
      width: 100%;
      display: flex;
      flex-direction: column;
      align-self: flex-end;

      .messages__author {
        align-self: flex-end;

      }

      .messages__date {
        align-self: flex-end;
      }

      .messages__message {
        align-self: flex-end;
        background: $message-light-color;
        color: $black-color;
      }
    }
  }
}

.background_hearts {
  background-image: url("../../../assets/background/hearts.jpg");
  background-size: 35%;
}

.background_crosses {
  background-image: url("../../../assets/background/crosses.jpg");
  background-size: 45%;
}

.background_leafs {
  background-image: url("../../../assets/background/leafs.jpg");
  background-size: 30%;
}

.background_hexals {
  background-image: url("../../../assets/background/hexals.jpg");
  background-size: 50%;
}

.background_triangles {
  background-image: url("../../../assets/background/triangles.jpg");
  background-size: 25%;
}

.background_papyrus {
  background-image: url("../../../assets/background/papyrus.jpg");
  background-size: 40%;
}

.background_morocco {
  background-image: url("../../../assets/background/morocco.jpg");
  background-size: 40%;
}
