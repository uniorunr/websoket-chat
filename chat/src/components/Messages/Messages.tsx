import React, { Component } from 'react';
import { StoreState } from '../../types';
import { connect } from 'react-redux';
import './Messages.scss';

class Messages extends Component {
  private messagesEnd: React.RefObject<HTMLDivElement>;

  constructor(props: any) {
    super(props);
    this.messagesEnd = React.createRef();
  }

  scrollToBottom = () => {
    const anchor = this.messagesEnd.current;
    if (anchor) anchor.scrollIntoView();
  };

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  render() {
    const { messages }: any = this.props;

    return (
      <div className="messages">
        <div className="messages__content">
          {messages.map((message: MessageObject) => (
            <div key={message.id}>
              <p className="messages__author"> {message.from}</p>
              <p className="messages__message">{message.message}</p>
            </div>
          ))}
          <div
            style={{ float: 'left', clear: 'both' }}
            ref={this.messagesEnd}
          />
        </div>
      </div>
    );
  }
}

interface MessageObject {
  id: string;
  message: string;
  from: string;
}

const mapStateToProps = (state: StoreState) => ({
  messages: state.messages
});

export default connect(mapStateToProps)(Messages);
