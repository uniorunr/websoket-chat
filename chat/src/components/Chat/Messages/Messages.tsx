import React, { Component } from 'react';
import { StoreState } from '../../../types';
import { connect } from 'react-redux';
import './Messages.scss';
import { formatDate } from './utils';

class Messages extends Component {
  private messagesEnd: React.RefObject<HTMLDivElement> = React.createRef();

  scrollToBottom = () => {
    const anchor = this.messagesEnd.current;
    if (anchor) anchor.scrollIntoView();
  };

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  render() {
    const { messages }: any = this.props;

    return (
      <div className="messages">
        <div className="messages__content">
          {messages.map((message: MessageObject) => (
            <div key={message.id}>
              <p className="messages__author">{message.from}</p>
              <p className="messages__message">{message.message}</p>
              <p className="messages__date">{formatDate(message.time)}</p>
            </div>
          ))}
          <div ref={this.messagesEnd} />
        </div>
      </div>
    );
  }
}

interface MessageObject {
  id: string;
  message: string;
  from: string;
  time: number;
}

const mapStateToProps = (state: StoreState) => ({
  messages: state.messages
});

export default connect(mapStateToProps)(Messages);
