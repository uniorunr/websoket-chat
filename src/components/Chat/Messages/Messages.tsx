import React, { Component } from 'react';
import { StoreState } from '../../../types';
import { connect } from 'react-redux';
import './Messages.scss';
import { formatDate } from './utils';

class Messages extends Component<MessagesProps> {
  private messagesEnd: React.RefObject<HTMLDivElement> = React.createRef();
  private container: React.RefObject<HTMLDivElement> = React.createRef();

  state = {
    scrollToBottom: true
  };

  scrollToBottom = () => {
    const { scrollToBottom } = this.state;
    if (scrollToBottom) {
      const anchor = this.messagesEnd.current;
      if (anchor) anchor.scrollIntoView();
    }
  };

  handleScroll = () => {
    if (this.container.current) {
      const scrollOffsetBottom =
        this.container.current.scrollHeight -
        this.container.current.offsetHeight;
      this.setState({
        scrollToBottom: scrollOffsetBottom === this.container.current.scrollTop
      });
    }
  };

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  render() {
    const { messages, userName, activeBackgroundId } = this.props;

    return (
      <div className={`messages background_${activeBackgroundId}`}>
        <div
          className="messages__content"
          ref={this.container}
          onScroll={this.handleScroll}
        >
          {messages.map((message: MessageObject) => (
            <div
              key={message.id}
              className={userName === message.from ? 'messages__own' : ''}
            >
              <p className="messages__author">{message.from}</p>
              <p className="messages__message">{message.message}</p>
              <p className="messages__date">{formatDate(message.time)}</p>
            </div>
          ))}
          <div ref={this.messagesEnd} />
        </div>
      </div>
    );
  }
}

interface MessagesProps {
  messages: never[];
  userName: string;
  activeBackgroundId: string;
}

interface MessageObject {
  id: string;
  message: string;
  from: string;
  time: number;
}

const mapStateToProps = (state: StoreState) => ({
  messages: state.messages,
  userName: state.userName,
  activeBackgroundId: state.settings.activeBackgroundId
});

export default connect(mapStateToProps)(Messages);
