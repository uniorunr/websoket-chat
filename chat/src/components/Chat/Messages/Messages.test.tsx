import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../../../store/mainStore';
import Messages from './Messages';

it('Messages component renders correctly', () => {
  const app = shallow(
      <Provider store={store}>
        <Messages />
      </Provider>
  );
  expect(app).toMatchSnapshot();
});
