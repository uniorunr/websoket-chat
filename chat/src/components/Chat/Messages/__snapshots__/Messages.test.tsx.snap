// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages component renders correctly 1`] = `ShallowWrapper {}`;
